tesseract-core/src/dataframe.rs

```diff
@@ -20,35 +21,65 @@ impl DataFrame {
-                ColumnData::NullableUInt32(ref ns) => ns.len(),
-                ColumnData::NullableUInt64(ref ns) => ns.len(),
-                ColumnData::NullableFloat32(ref ns) => ns.len(),
-                ColumnData::NullableFloat64(ref ns) => ns.len(),
-                ColumnData::NullableText(ref ss) => ss.len(),
-            }
-        } else {
-            0
+        match self.columns.get(0) {
+            Some(col) => col.len(),
+            None => 0,
         }
     }
+
+    pub fn drain_filter<P>(mut self, predicate: P) -> Self
+    where
+        P: Fn(HashMap<&str, Datum>, usize) -> bool,
+    {
+        let col_amount = self.columns.len();
+
+        let mut i = 0;
+        while i < self.len() {
+            let mut datum: HashMap<&str, Datum> = HashMap::with_capacity(col_amount);
+
+            for column in &self.columns {
+                let value: Datum = match &column.column_data {
+                    ColumnData::Int8(d) => Datum::Int8(d[i]),
+                    ColumnData::Int16(d) => Datum::Int16(d[i]),
+                    ColumnData::Int32(d) => Datum::Int32(d[i]),
+                    ColumnData::Int64(d) => Datum::Int64(d[i]),
+                    ColumnData::UInt8(d) => Datum::UInt8(d[i]),
+                    ColumnData::UInt16(d) => Datum::UInt16(d[i]),
+                    ColumnData::UInt32(d) => Datum::UInt32(d[i]),
+                    ColumnData::UInt64(d) => Datum::UInt64(d[i]),
+                    ColumnData::Float32(d) => Datum::Float32(d[i]),
+                    ColumnData::Float64(d) => Datum::Float64(d[i]),
+                    ColumnData::Text(d) => Datum::Text(d[i].to_string()),
+                    ColumnData::NullableInt8(d) => Datum::NullableInt8(d[i]),
+                    ColumnData::NullableInt16(d) => Datum::NullableInt16(d[i]),
+                    ColumnData::NullableInt32(d) => Datum::NullableInt32(d[i]),
+                    ColumnData::NullableInt64(d) => Datum::NullableInt64(d[i]),
+                    ColumnData::NullableUInt8(d) => Datum::NullableUInt8(d[i]),
+                    ColumnData::NullableUInt16(d) => Datum::NullableUInt16(d[i]),
+                    ColumnData::NullableUInt32(d) => Datum::NullableUInt32(d[i]),
+                    ColumnData::NullableUInt64(d) => Datum::NullableUInt64(d[i]),
+                    ColumnData::NullableFloat32(d) => Datum::NullableFloat32(d[i]),
+                    ColumnData::NullableFloat64(d) => Datum::NullableFloat64(d[i]),
+                    ColumnData::NullableText(d) => match &d[i] {
+                        Some(nt) => Datum::NullableText(Some(nt.to_string())),
+                        None => Datum::NullableText(None),
+                    },
+                };
+
+                datum.insert(&column.name, value);
+            }
+
+            if !predicate(datum, i) {
+                for column in self.columns.iter_mut() {
+                    column.remove(i);
```

Comment:
Don't know how big the database is, but repeatedly calling `remove` could be expensive.